lib/supavisor_web/controllers/tenant_controller.ex

```diff
@@ -82,34 +89,68 @@ defmodule SupavisorWeb.TenantController do
     }
   )
 
-  def update(conn, %{"external_id" => id, "tenant" => tenant_params}) do
-    case H.check_creds_get_ver(tenant_params) do
-      {:error, reason} ->
+  # conver cert to pem format
+  def update(conn, %{
+        "external_id" => id,
+        "tenant" => %{"upstream_tls_ca" => "-----BEGIN" <> _ = upstream_tls_ca} = tenant_params
+      }) do
+    case H.cert_to_bin(upstream_tls_ca) do
+      {:ok, bin} ->
+        update(conn, %{
+          "external_id" => id,
+          "tenant" => %{tenant_params | "upstream_tls_ca" => bin}
+        })
+
+      {:error, realson} ->
         conn
         |> put_status(400)
-        |> render("error.json", error: reason)
-
-      {:ok, pg_version} ->
-        tenant_params =
-          Map.put(tenant_params, "default_parameter_status", %{"server_version" => pg_version})
-
-        case Tenants.get_tenant_by_external_id(id) do
-          nil ->
-            create(conn, %{"tenant" => Map.put(tenant_params, "external_id", id)})
+        |> render("error.json",
+          error: "Invalid 'upstream_tls_ca' certificate, reason: #{inspect(realson)}"
+        )
+    end
+  end
 
-          tenant ->
-            tenant = Repo.preload(tenant, :users)
+  def update(conn, %{"external_id" => id, "tenant" => params}) do
+    cert = H.upstream_cert(params["upstream_tls_ca"])
 
-            with {:ok, %TenantModel{} = tenant} <- Tenants.update_tenant(tenant, tenant_params) do
-              for user <- tenant.users do
-                Supavisor.stop(tenant.external_id, user.db_user_alias)
-                |> then(&"Stop #{user.db_user_alias}.#{tenant.external_id}: #{inspect(&1)}")
-                |> Logger.warning()
+    if params["upstream_ssl"] && params["upstream_verify"] == "peer" && !cert do
+      conn
+      |> put_status(400)
+      |> render("error.json",
+        error: "Invalid 'upstream_verify' value, 'peer' is not allowed without certificate"
+      )
+    else
+      case H.check_creds_get_ver(params) do
+        {:error, reason} ->
+          conn
+          |> put_status(400)
+          |> render("error.json", error: reason)
+
+        {:ok, pg_version} ->
+          params =
+            Map.put(params, "default_parameter_status", %{
+              "server_version" => pg_version
+            })
+
+          case Tenants.get_tenant_by_external_id(id) do
+            nil ->
+              create(conn, %{"tenant" => Map.put(params, "external_id", id)})
+
+            tenant ->
+              tenant = Repo.preload(tenant, :users)
+
+              with {:ok, %TenantModel{} = tenant} <-
+                     Tenants.update_tenant(tenant, params) do
+                for user <- tenant.users do
+                  Supavisor.stop(tenant.external_id, user.db_user_alias)
+                  |> then(&"Stop #{user.db_user_alias}.#{tenant.external_id}: #{inspect(&1)}")
+                  |> Logger.warning()
```

Comment:
nitpick: this is subjective but in this particular case I'd really consider breaking up the pipe:

```suggestion
                  result = Supavisor.stop(tenant.external_id, user.db_user_alias)
                  Logger.warning("Stop #{user.db_user_alias}.#{tenant.external_id}: #{inspect(result)}")
```